{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal of this project is to re-examine the meta4 metagenomic data for mutations of interest. This will first be done by looking at the raw reads re-aligned to isolate genomes. Genes from interesting protein families and pathways will then be pulled out to look for a change in mutations over time as the microbial community underwent generations of passaging."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# fastq files\n",
    "\n",
    "**to view zipped fastq.gz files on command line:**\n",
    "\n",
    "`gzip -cd 8776.4.111692.CGATGT.fastq.gz | more`\n",
    ">@HISEQ09:160:C6FEEANXX:4:1101:1469:2228 1:N:0:CGATGT\n",
    ">GCATACCTGCAATCCTATTCTGAATATGGACAAAACATATTCGATCGTTATGTGACCTATGCCGACTATTGGATACAGGACACCAATTATCGCGATCCTGATACCGGCGAAATGTTCGATAGAGGCGCCTTAAATAATGAACTGGAAAAAA\n",
    ">+\n",
    ">CCCCCGGGGGGGGBFGGGEGGGGCDGGFGGGGGGGGGGGGGGGCC@DGGGGEGGCFGGEGG1EDACEGGG0FGGGEGGG00FGG>GGGGGGFFGGG:8CGGGDG>BGGDGGGGGGGEGBEGGEGGBGD.@>GGBGBGG=BGGG..335EBD\n",
    ">@HISEQ09:160:C6FEEANXX:4:1101:1469:2228 2:N:0:CGATGT\n",
    ">TTGGCATTAAATGAAATCACTGGCAACAAATCTTCAGTGGTTGAGAACATTTTCTTTTCACTGACTACTCGCAGTTTTTCCTAACTGGTCCAGACAGGATTCTTGCCTTCCTGTTTCGAACGGGACCGTAAAACGAAATTGACGATTTCAT\n",
    ">+\n",
    ">CBBBB@GGGGGGGGGGGGGGGGGGGGDFGEGGGGGGFGC>EFG@F>FGFGGGGGDCGGGG1:BGGGGGG1=EGGGG>FGGDD00=FD0>C0FCCFBGGBGGEDGGCGCEF0;FFGGG<008CBB..9=CDB88DC668@GGBBC#######\n",
    "\n",
    "**count number of lines in fastq.gz file:**\n",
    "\n",
    "`gzip -cd 8776.4.111692.CGATGT.fastq.gz | wc -l`\n",
    ">130259312\n",
    "\n",
    "**count number of reads in fastq.gz file:**\n",
    "\n",
    "`gzip -cd 8776.4.111692.CGATGT.fastq.gz | grep -c '@HISEQ'`\n",
    ">32564828"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bwa & bwa mem\n",
    "\n",
    "http://bio-bwa.sourceforge.net/bwa.shtml\n",
    "\n",
    "#### SYNOPSIS\n",
    "\n",
    ">bwa index ref.fa\n",
    "\n",
    ">bwa mem ref.fa reads.fq > aln-se.sam\n",
    "\n",
    ">bwa mem ref.fa read1.fq read2.fq > aln-pe.sam\n",
    "\n",
    ">bwa aln ref.fa short_read.fq > aln_sa.sai\n",
    "\n",
    ">bwa samse ref.fa aln_sa.sai short_read.fq > aln-se.sam\n",
    "\n",
    ">bwa sampe ref.fa aln_sa1.sai aln_sa2.sai read1.fq read2.fq > aln-pe.sam\n",
    "\n",
    ">bwa bwasw ref.fa long_read.fq > aln.sam\n",
    "\n",
    "#### Scripting example\n",
    "\n",
    "https://pyjip.readthedocs.io/en/latest/examples/bwa.html\n",
    "\n",
    "#### Python binding to bwa mem\n",
    "\n",
    "https://github.com/nanoporetech/bwapy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# samtools\n",
    "\n",
    "http://samtools.sourceforge.net/\n",
    "    \n",
    "Introduction\n",
    "\n",
    "SAM (Sequence Alignment/Map) format is a generic format for storing large nucleotide sequence alignments. SAM aims to be a format that:\n",
    "\n",
    "* Is flexible enough to store all the alignment information generated by various alignment programs;\n",
    "* Is simple enough to be easily generated by alignment programs or converted from existing alignment formats;\n",
    "* Is compact in file size;\n",
    "* Allows most of operations on the alignment to work on a stream without loading the whole alignment into memory;\n",
    "* Allows the file to be indexed by genomic position to efficiently retrieve all reads aligning to a locus.\n",
    "\n",
    "SAM Tools provide various utilities for manipulating alignments in the SAM format, including sorting, merging, indexing and generating alignments in a per-position format.\n",
    "\n",
    "SAMtools is hosted by GitHub. The project page is here. The source code releases are available from the download page. You can check out the most recent source code with:\n",
    "\n",
    "`git clone git://github.com/samtools/samtools.git`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# seqtk\n",
    "\n",
    "tool for fasta & fastq file manipulation\n",
    "\n",
    "**written in C/C++**\n",
    "\n",
    "from github:\n",
    "\n",
    ">Seqtk is a fast and lightweight tool for processing sequences in the FASTA or FASTQ format. It seamlessly parses both FASTA and FASTQ files which can also be optionally compressed by gzip.\n",
    "\n",
    "\n",
    "https://wiki.gacrc.uga.edu/wiki/Seqtk\n",
    "\n",
    "https://github.com/lh3/seqtk\n",
    "\n",
    "Subsample 10000 read pairs from two large paired FASTQ files (remember to use the same random seed to keep pairing):\n",
    "```\n",
    "  seqtk sample -s100 read1.fq 10000 > sub1.fq\n",
    "  seqtk sample -s100 read2.fq 10000 > sub2.fq\n",
    "```\n",
    "\n",
    "#### My solution:\n",
    "\n",
    "`gzip -cd 8776.4.111692.CGATGT.fastq.gz | seqtk sample -s42 - 100000 > sandbox.fastq`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bedtools\n",
    "\n",
    "https://bedtools.readthedocs.io/en/latest/\n",
    "\n",
    "**enables all kinds of genome alignment statistics - look at this tutorial!!!**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gff from genbank & fasta\n",
    "\n",
    "\n",
    "<https://biopython.org/wiki/GFF_Parsing>\n",
    "\n",
    "<https://github.com/chapmanb/bcbb/tree/master/gff>\n",
    "\n",
    "<https://github.com/daler/gffutils>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check genome annotations\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# alignment visualization tools\n",
    "\n",
    "> samtools tview alignment.bam\n",
    "\n",
    "```bash\n",
    "[bam_header_read] EOF marker is absent. The input is probably truncated.\n",
    "[bam_index_load] fail to load BAM index.\n",
    "Cannot read index for 'alignments.bam.unsorted'.\n",
    "```\n",
    "\n",
    "genome browsers, yes\n",
    "\n",
    "others?\n",
    "\n",
    "#### igv\n",
    "\n",
    "https://igv.org/\n",
    "\n",
    "\n",
    "**do this tutorial!!!**\n",
    "\n",
    "http://fullstackdatascientist.io/15/03/2016/genomic-data-visualization-using-python/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# docker container\n",
    "\n",
    "_for building a reproducible computation environment_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# questions for Dave:\n",
    "\n",
    "- why did it take so long to copy some files from my waffle mount directory to a local directory?\n",
    "    - why did my terminal freezes up when I'm trying to copy things?\n",
    "- good alignment visualization\n",
    "    - good comparison of alignment visualization tools\n",
    "- talk through gene selection process\n",
    "    - Kegg I think\n",
    "    - Joseph can you highlight a list of 3-5 genes to look at first for messing around?\n",
    "- for tool building: select genes before or after alignment?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## variant calling:\n",
    "\n",
    "http://www.gettinggeneticsdone.com/2014/03/visualize-coverage-exome-targeted-ngs-bedtools.html\n",
    "\n",
    "https://wikis.utexas.edu/display/bioiteam/Variant+calling+using+SAMtools\n",
    "\n",
    "https://samtools.github.io/bcftools/howtos/variant-calling.html\n",
    "\n",
    "http://biobits.org/samtools_primer.html#IdentifyingGenomicVariants\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
